Range performance fixes

wjt opened the pull request:
I'm using Annotator on pages with really large tables, highlighting potentially thousands of rows of text. Selecting any amount of text which spans more than one row was extremely expensive. `NormalizedRange#limit` and `#getTextNodes` were the main culprits according to Chrome's profiler. These patches make them an order of magnitude faster for me.

Happy to write more tests if you're not convinced that this new logic is right! Also, the existing code is inconsistent in whether it uses `$.contains` or `util.contains`; shout if there's some hidden reason to use one or the other that I've not seen.

Agh, while writing this pull request I've noticed that I introduced a use of `childNodes` in `#limit` -- based on the comment in `#getTextNodes` I guess I need to remove that for IE9's sake.

wjt committed:
Test NormalizedRange#limit more thoroughly

Previously, only the left-hand side was tested; and grandchildren of the
bounds were not tested at all. With [] indicating a text node:

    <h1>[My Heading]</h1>
    <p>
      [My paragraph]
      <span>
        [ conti]
        [nues]
      </span>
    <p>
    <p>[Another paragraph begins]</p>

I wrote an implemetation where, given a range ending in the second
paragraph, limiting it to the first paragraph would incorrectly update
the `@end` to ` conti` not `nues`. The old test case did not catch it.

wjt committed:
NormalizedRange#limit: use TreeWalker

Traversing the whole tree is unnecessary -- and costly if the
@commonAncestor is <body> and you have many nodes in your document.

All we actually need to do is update @start and @end to point at the
first and last text nodes in the bounds, if they currently fall outside
it.

* master: 1.61s ±2.26% (6 runs sampled) -- but this is a little unfair
  since it was implemented in terms of a very slow getTextNodes call
* this implementation: 0.01ms ±2.91% (85 runs sampled)

wjt committed:
getTextNodes: walk the tree using TreeWalker

The old implementation was really expensive on a range covering many
rows of table. I think it's because an array was being allocated at
every level of the tree, then flattened out into the level above.

Instead here we just do a depth-first search for text nodes and build up
a single array as we go.  In a little benchmark with a range covering
every cell of a 5000×3 table:

* master: 1.57s ±1.14% (6 runs sampled)
* TreeWalker: 17.5ms ±0.74% (72 runs sampled)

tilgovi commented on src/range.coffee:
```diff
@@ -189,13 +189,33 @@ class Range.NormalizedRange
   limit: (bounds) ->
-    nodes = $.grep this.textNodes(), (node) ->
-      node.parentNode == bounds or $.contains(bounds, node.parentNode)
+    if @commonAncestor == bounds or $.contains(bounds, @commonAncestor)
```
I would make this `===` for clarity.

wjt replied:
I've not written CoffeeScript before yesterday, but [the docs](http://coffeescript.org/#operators) suggest that `==` in CoffeeScript compiles to `===` in JS.

tilgovi replied:
I'd forgotten this project was still coffee!

Brain splat.

tilgovi commented on src/util.coffee:
```diff
@@ -60,27 +60,28 @@ Util.getLastTextNodeUpTo = (n) ->
-  getTextNodes = (node) ->
-    if node and node.nodeType != Util.NodeTypes.TEXT_NODE
-      nodes = []
+  getTextNodes = (root) ->
```
I wonder how this compares to:

```coffeescript
walker = getGlobal().document.createTreeWalker(root, NodeFilter.SHOW_TEXT)
nodes = (node while node = walker.nextNode())
```

wjt replied:
I will write a little benchmark and investigate.

wjt commented:
To no-one's great surprise, using `TreeWalker` is a bit faster (or no slower) and more concise. I didn't know about it, thanks! I also found a bug in my open-coded implementation of `#limit` – the tests now cover that.

wjt commented:
(Let me know if you would like the intermediate patches squashed away.))

tilgovi commented:
That's excellent. Squash away and I'll be glad to merge it!

tilgovi commented:
If you could also remove the flatten function, which is no longer used now, that would be excellent.

wjt committed:
Remove Util#flatten: it's now unused

wjt committed:
Stop using Util.getGlobal() to get document

Per openannotation/annotator#527 . We actually don't need to use global
at all -- we can get the document from the element we're about to walk.

wjt commented:
How's that?

tilgovi commented:
Looks great :)

mwidner commented:
@tilgovi @wjt How challenging do you think it'd be to backport this to 1.2.x? I'm seeing the same performance issue there and can't upgrade to 2.0 yet. Thanks.

tilgovi commented:
@mwidner there's a v1.2.x branch on annotator so we have a place to backport fixes and make maintenance releases. Do you think you could extract this into a PR there?

tilgovi commented:
It shouldn't be too hard because when this change happened the files were still coffee and still had the same names and paths, so it might be as simple as taking a copy of the annotator repo, branching off v1.2.x, adding this repo as a remote and cherry-picking these changes.

mwidner commented:
Done! Thanks for the pointers.
